### ISSUE TITLE
podman-docker's "docker build --pull=false ..." behaves differently than in docker

### ISSUE THREAD

**Issue by vorburger** [containers/buildah#1675]
**Description**

It appears that a `docker build --pull=false ...` behaves differently in `podman-docker` than in "real" docker; see https://github.com/kubernetes/kubernetes/issues/79048 for an analysis.

**Steps to reproduce the issue:**

1. Fedora 30 Server
1. sudo dnf install podman-docker
1. git clone https://github.com/kubernetes/kubernetes.git ; cd kubernetes
1. build/run.sh make kubectl

fails on Buildah in Podman saying `Error: error creating build container: no such image "containers-storage:[overlay@/home/vorburger/.local/share/containers/storage+/tmp/1000:overlay.mount_program=/usr/bin/fuse-overlayfs]k8s.gcr.io/kube-cross:v1.12.5-1"` but (apparently) works in regular docker.

**Output of `podman version`:**

```
$ podman version
Version:            1.3.1
RemoteAPI Version:  1
Go Version:         go1.12.2
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
$ podman info --debug
debug:
  compiler: gc
  git commit: ""
  go version: go1.12.2
  podman version: 1.3.1
host:
  BuildahVersion: 1.8.2
  Conmon:
    package: podman-1.3.1-1.git7210727.fc30.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: c9a4c48d1bff85033b7fc9b62d25961dd5048689'
  Distribution:
    distribution: fedora
    version: "30"
  MemFree: 13008052224
  MemTotal: 33585057792
  OCIRuntime:
    package: runc-1.0.0-93.dev.gitb9b6cc6.fc30.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc8+dev
      commit: e3b4c1108f7d1bf0d09ab612ea09927d9b59b4e3
      spec: 1.0.1-dev
  SwapFree: 16859000832
  SwapTotal: 16859000832
  arch: amd64
  cpus: 8
  hostname: toby
  kernel: 5.1.6-300.fc30.x86_64
  os: linux
  rootless: true
  uptime: 153h 43m 38.99s (Approximately 6.38 days)
registries:
  blocked: null
  insecure: null
  search:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ConfigFile: /home/vorburger/.config/containers/storage.conf
  ContainerStore:
    number: 1
  GraphDriverName: overlay
  GraphOptions:
  - overlay.mount_program=/usr/bin/fuse-overlayfs
  GraphRoot: /home/vorburger/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 3
  RunRoot: /tmp/1000
  VolumePath: /home/vorburger/.local/share/containers/storage/volumes
```

**Output of `cat /etc/*release`:**

```
Fedora release 30 (Thirty)
NAME=Fedora
VERSION="30 (Server Edition)"
ID=fedora
VERSION_ID=30
VERSION_CODENAME=""
PLATFORM_ID="platform:f30"
PRETTY_NAME="Fedora 30 (Server Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:30"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f30/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=30
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=30
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Server Edition"
VARIANT_ID=server
Fedora release 30 (Thirty)
Fedora release 30 (Thirty)
```

**Output of `uname -a`:**

```
Linux toby 5.1.6-300.fc30.x86_64 #1 SMP Fri May 31 17:43:23 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev,metacopy=on"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base.
# device.
# mkfsarg = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver 
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

# If specified, use OSTree to deduplicate files with the overlay backend
ostree_repo = ""

# Set to skip a PRIVATE bind mount on the storage home directory.  Only supported by
# certain container storage drivers
skip_mount_home = "false"
```

**Comment by rhatdan** [comment]
@vorburger What is the difference.  This looks like `podman build --pull=false`  Does not pull the image even if the image does not exist, does `docker build --pull=false` pull the image if it does not exists locally?

**Comment by vorburger** [comment]
> @vorburger What is the difference. This looks like `podman build --pull=false` Does not pull the image even if the image does not exist, does `docker build --pull=false` pull the image if it does not exists locally?

I'm not 100% sure what the exact behaviour of `docker build --pull=false` is if the parent `FROM` image does not exist (and will admit that I haven't actually tried it out myself), but basically yes, over in https://github.com/kubernetes/kubernetes/issues/79048 your fellow red hat @cblecker said the following (and based on this refused, rightfully, to change `--pull=false` to `--pull=true` in some K8s scripts, because "it works in docker so it's podman build's problem"), hear:

> Looks like it has to do with a difference in flag meanings between docker and podman. `docker build --pull=false` seems to be the equivalent of a `podman build --pull-always=false`. `podman build --pull=false` has a different meaning and functionality.
> 
> I would suggest opening up a bug report against libpod.. if they're going to provide an emulation layer, it should probably associate the right flags.

@cblecker perhaps you'd like to chime in here?

**Comment by rhatdan** [comment]
@nalind WDYT?
`buildah bud` has --pull and --pull-always (`buildah bud --pull-always` is the equivalent of `docker build --pull`)
The problem is when we vendor this into `podman build`, users expect it's --pull to be the same as docker build.

**Comment by rhatdan** [comment]
In the `podman build` man page, we describe the behaviour:
```
**--pull**

When the flag is enabled, attempt to pull the latest image from the registries
listed in registries.conf if a local image does not exist or the image is newer
than the one in storage. Raise an error if the image is not in any listed
registry and is not present locally.

If the flag is disabled (with *--pull=false*), do not pull the image from the
registry, use only the local version. Raise an error if the image is not
present locally.

Defaults to *true*.

**--pull-always**

Pull the image from the first registry it is found in as listed in registries.conf.
Raise an error if not found in the registries, even if the image is present locally.
```

Does the tool currently follow this behaviour?

**Comment by rhatdan** [comment]
`docker build --pull` is documented as:
```
       --pull=true|false
          Always attempt to pull a newer version of the image. The default is
       false.
```

**Comment by cblecker** [comment]
Yup, these are accurate. podman/buildah offer two different, more explicit settings than the docker cli does.. but the issue is when using the podman-docker compatibility/emulation layer, `--pull` has a different meaning than the docker CLI.

**Comment by rhatdan** [comment]
I think what we should do is eliminate make --pull-always == --pull for Podman and not support --pull at all.

The problem then would be `buildah bud --pull` would mean a different thing the `podman build --pull`

**Comment by TomSweeneyRedHat** [comment]
Fixed in Buildah, will be fixed in the next version of  Podman after Buildah is vendored.

**Comment by TomSweeneyRedHat** [comment]
Fixed with #1959 which has since been vendored into Podman.
